gradle/publish-maven.gradle

```diff
@@ -1,105 +1,91 @@
 }
-
-//upload to sonatype configuration
-uploadArchives {
+//define the repositories on global level, not per project, or you get many of them
+publishing {
   repositories {
-    mavenDeployer {
-      if(gradle.ext.shouldSign) {
-        beforeDeployment { MavenDeployment deployment -> signing.signPom(deployment) }
-      }
+    maven {
+      name = "ossrh"
+      //  if(gradle.ext.shouldSign) {
+      //    beforeDeployment { MavenDeployment deployment -> signing.signPom(deployment) }
+      //  }
+      // change URLs to point to your repos, e.g. http://my.org/repo
+      def releasesRepoUrl = 'https://oss.sonatype.org/service/local/staging/deploy/maven2/'
+      def snapshotsRepoUrl = 'ttps://oss.sonatype.org/content/repositories/snapshots/'
+      url = version.endsWith('SNAPSHOT') ? snapshotsRepoUrl : releasesRepoUrl
 
-      repository(url: "https://oss.sonatype.org/service/local/staging/deploy/maven2/") {
-        authentication(userName: gradle.ext.ossrhUsernameSafe, password: gradle.ext.ossrhPasswordSafe)
-      }
-
-      snapshotRepository(url: "https://oss.sonatype.org/content/repositories/snapshots/") {
-        authentication(userName: gradle.ext.ossrhUsernameSafe, password: gradle.ext.ossrhPasswordSafe)
-      }
-      customizePom(pom, project)
-      pom.whenConfigured { generatedPom ->
-        if(gradle.ext.isParallel) {
-          logger.error('build error: this build file cannot yet upload to maven central with parallel builds enabled')
-          throw new InvalidUserDataException('build error: this build file cannot yet upload to maven central with parallel builds enabled')
-        }
+      credentials {
+        username = project.hasProperty('ossrhUsername') ? ossrhUsername : "Unknown user"
+        password = project.hasProperty('ossrhPassword') ? ossrhPassword : "Unknown password"
       }
     }
   }
 }
 
 
+//now uploading will have to be configured on a per subproject basis
+subprojects {
 
-def customizePom(pom, gradleProject) {
-  pom.whenConfigured { generatedPom ->
-
-    //description is required for publishing, so enforce it
-    if(gradleProject.description == null || gradleProject.description == '') {
-      logger.error('build error: description required for release to maven central, but not set for subproject ' + gradleProject.name)
-      throw new InvalidUserDataException('build error: description required for release to maven central, but not set for subproject ' + gradleProject.name)
-    }
+  publishing {
 
-    // eliminate test-scoped dependencies (no need in maven central poms)
-    generatedPom.dependencies.removeAll { dep ->
-      dep.scope == "test"
-    }
+    publications {
+      "$project.name"(MavenPublication) {
+        afterEvaluate {
+          from components.java
+          artifact sourcesJar
+          artifact javadocJar
+          artifactId = project.name
+          pom {
+            name = project.name
+            description = project.description
+          }
+        }
```

Comment:
This `afterEvaluate` block is actually required?

Would it be possible to use
```
java {
    withJavadocJar()
    withSourcesJar()
}
```
instead of these custom gradle tasks?

Reply:
Ah, good suggestion. This was from a rather old gradle build-file, this is better indeed.